### PR TITLE
Add getPublicKey by cuid

### LINKED ISSUE #75: Use cuid to create a non-PII user identity slug
Part of  [FrostAPI #243](https://github.com/poetapp/frost-api/issues/243)

### PR DESCRIPTION
**Pull request by krobi64**
[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

### Release
 Semantic release is enabled for this repository. Make sure you follow the right commit message convention. 
 We're using semantic-release's default — [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

## Description of Changes

Resolves #75
Depends on https://github.com/poetapp/frost-api/pull/299 for actual use

**feat(*)**

Introduces support for an automatically-generated unique identifier for each account based upon the https://github.com/ericelliott/cuid library. 

**feat(*)**

Adds the getPublicKey method for retrieving a users public keys when supplied a cuid. This is required if we are going to use jsonld-signing.

### CODE REVIEW

**Comment by warrenv** [comment]
I'm should be seeing cuid in package.json right?

**Comment by krobi64** [comment]
frost-api will generate the actual cuid upon calling create here. The actual library is installed there.

File: src/Frost.ts [hunk 1]
```diff
@@ -341,4 +344,24 @@ export class Frost {
...
+    }
+
+    const request = fetch(`${this.host}${Path.PUBLIC_KEYS}/${cuid || this.cuid}`, options)
+
+    return Promise.race([request, this.timeoutPromise()])
+      .then(async (value: any) => {
+        if (value.ok) return await value.json()
+        throw await value.text()
+      })
+      .catch(e => {
```
Comment (lautarodragan) [inline src/Frost.ts]: catch(e => { throw e}) is redundant

File: src/Frost.ts [hunk 2]
```diff
@@ -341,4 +344,24 @@ export class Frost {
         throw e
       })
   }
+
+  getPublicKeys(cuid?: string): Promise<{ readonly publicKey: string }> {
```
Comment (lautarodragan) [inline src/Frost.ts]: Is this endpoint available in Frost API already?

Can't find it in `master`

**Comment by krobi64** [comment]
I'll try and extracate it.

File: src/Frost.ts (hunk 2 above)
Comment (lautarodragan) [inline src/Frost.ts]: Maybe name this PR as to refer to this new endpoint being added as the main thing rather than adding cuid?

**Comment by lautarodragan** [comment]
@krobi64 @warrenv actually the package-lock should see a change for version changes. `npm version minor` should do the trick.

**Comment by lautarodragan** [comment]
@krobi64 @warrenv scratch that, if we're using semantic release in this repo, we shouldn't be manually updating the version tag altogether.

@wzalazar can you confirm/help? Not sure what the package.json should look like.

**Comment by wzalazar** [comment]
@wzalazar can you confirm/help? Not sure what the package.json should look like.

it not necessary, but `semantic-realease` doesn't change the version in the `package.json` in `git` by default

> Why is the package.json’s version not updated in my repository?
> semantic-release takes care of updating the package.json’s version before publishing to npm.
> 
> By default, only the published package will contains the version, which is the only place where it is really required, but the updated package.json will not be pushed to the Git repository
> 
> However, the @semantic-release/git plugin can be used to push the updated package.json as well as other files to the Git repository.
> 

at moment we don't implement the update the package.json over git

File: src/Frost.ts (hunk 2 above)
Comment (krobi64) [reply src/Frost.ts]: @lautarodragan  Which comes first the client or the api? The api seems to rely on the client, so I presumed this came first.

**Comment by ericelliott** [comment]
## PR Review Checklist


* [ ] Tested changes manually
* [x] Checked accidental architectural/style changes
* [x] Reviewed entire diff
* [ ] Unit tests
* [ ] Documentation
* [N/A] Filenames and locations

### PR Reviewer Comments

Do we need to add tests for this?
Are there any interface docs that need updating?

**Comment by krobi64** [comment]
@ericelliott I didn't add anything to the README because we don't have documentation for the `create` method or the `getProfile` method. If those should be there, I will happily do that in addition to adding this method.

**Comment by lautarodragan** [comment]
@krobi64 

> @lautarodragan Which comes first the client or the api? The api seems to rely on the client, so I presumed this came first.

The API should should only rely on the client for functional tests. I believe this is the case.

The client is just a simple wrapper over the API endpoints of Frost, so Frost API comes first.

File: src/Frost.ts [hunk 3]
```diff
@@ -331,14 +279,25 @@ export class Frost {
...
+      throw await value.text()
+    })
+  }
 
-        throw await value.text()
-      })
-      .catch(e => {
-        throw e
-      })
+  getPublicKeys(cuid?: string): Promise<[string]> {
```
Comment (lautarodragan) [inline src/Frost.ts]: `[string]` is [a tuple type](https://www.typescriptlang.org/docs/handbook/basic-types.html). 

It's declaring that this method always returns an array with a single element. 

Is this the intended behavior?

If we want to specify that it returns an array, the type annotation should be `Promise<ReadonlyArray<string>>` instead.

Comment (lautarodragan) [reply src/Frost.ts]: An array of strings, of any length. `string[]` is shorthand for `Array<string>`, but `ReadonlyArray` doesn't have a shorthand version.

Thinking again about it, though, it shouldn't return a ReadonlyArray, so `Promise<string[]>` is ok.

**Comment by lautarodragan** [comment]
@krobi64 continuing my last comment on which depends on which... I now realize that since Frost API uses Frost Client for the tests, the build will fail until the new version is released.

I'd rather the Frost Client always represent the `master` branch of Frost API (or a released version), but there's no real damage in releasing it "early", so let's go with that.

File: package.json [hunk 4]
```diff
@@ -1,9 +1,9 @@
 {
   "name": "@po.et/frost-client",
-  "version": "0.1.8",
+  "version": "0.2.0",
   "description": "Po.et - Frost client",
-  "main": "dist/Frost.js",
-  "types": "dist/Frost.d.ts",
+  "main": "dist/src/Frost.js",
```
[the remark this answers is not in the document]
Comment (krobi64) [reply package.json]: The change here was due to a change in `tsconfig.json`
From @wzalazar 
when was applied this change
 ```"include": [
    "./src/**/*"
  ],```
to
```"include": [
      "./src/**/*",
      "./src/**/*.test.ts",
      "./typings/**/*"
    ]``` 

So are you asking that I create a Frost.d.ts file?

**Comment by krobi64** [comment]
I will be closing this in favor of a new PR to abstract profiles from the user account.

**Comment by krobi64** [comment]
I will be closing this PR in favor of one that abstracts the profile from the user identity

**Pull request closed without merging**